{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"../OUTPUTS/antismash_finals/T_712_11/mg.assembly.merged.length_filtered/knownclusterblastoutput.txt\", 'r')\n",
    "#for Line in f:\n",
    " #       columns = Line.split('|')\n",
    "  #      BGC_id = columns[5].split('\\t')[1]\n",
    "   #     print(BGC_id)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: PFAM_domain\n",
      "location: [0:309](-)\n",
      "qualifiers:\n",
      "    Key: aSDomain, Value: ['LeuA_dimer']\n",
      "    Key: aSTool, Value: ['clusterhmmer']\n",
      "    Key: database, Value: ['34.0']\n",
      "    Key: db_xref, Value: ['PF08502.12']\n",
      "    Key: description, Value: ['LeuA allosteric (dimerisation) domain']\n",
      "    Key: detection, Value: ['hmmscan']\n",
      "    Key: domain_id, Value: ['clusterhmmer_JJJEIFBD_328428_0002']\n",
      "    Key: evalue, Value: ['5.40E-11']\n",
      "    Key: label, Value: ['JJJEIFBD_328428']\n",
      "    Key: locus_tag, Value: ['JJJEIFBD_328428']\n",
      "    Key: protein_end, Value: ['311']\n",
      "    Key: protein_start, Value: ['208']\n",
      "    Key: score, Value: ['42.6']\n",
      "    Key: tool, Value: ['antismash']\n",
      "    Key: translation, Value: ['EGADASFELLARRTLGTIPEFFGIDGFRVMIERRFDSLGRVKIVSEAVVKITIDGQTLMSVAEAEGPVNALDLALRKDFGKYQHEIDDLVLADFKVRILNGGT']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def get_cds_feature_with_qualifier_value(seq_record, name, value):\n",
    "    \"\"\"Function to look for CDS feature by annotation value in sequence record.\n",
    "    \n",
    "    e.g. You can use this for finding features by locus tag, gene ID, or protein ID.\n",
    "    \"\"\"\n",
    "    # Loop over the features\n",
    "    for feature in genome_record.features:\n",
    "        if feature.type == \"PFAM_domain\" and value in feature.qualifiers.get(name, []):\n",
    "            return feature\n",
    "    # Could not find it\n",
    "    return None\n",
    "\n",
    "genome_record = SeqIO.read(\"../OUTPUTS/antismash_finals/T_712_11/mg.assembly.merged.length_filtered/test_contig_240684.region001.gbk\", \"genbank\")\n",
    "cds_feature = get_cds_feature_with_qualifier_value(genome_record, \"locus_tag\", \"JJJEIFBD_328428\")\n",
    "\n",
    "print(cds_feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['test_contig_240684\\tantismash\\tPFAM_domain\\t1\\t309\\t.\\t-\\t0\\taSDomain=LeuA_dimer;aSTool=clusterhmmer;database=34.0;db_xref=PF08502.12;description=LeuA allosteric (dimerisation) domain;detection=hmmscan;evalue=5.40E-11;label=JJJEIFBD_328428;locus_tag=JJJEIFBD_328428;protein_end=311;protein_start=208;score=42.6;tool=antismash;', \"test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t1\\t933\\t.\\t-\\t0\\tgene_functions=['biosynthetic (rule-based-clusters) betalactone: HMGL-like', 'biosynthetic-additional (smcogs) SMCOG1271:2-isopropylmalate synthase (Score: 230.1; E-value: 8.7e-70)'];gene_kind=biosynthetic;inference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328428;phase=0;product=unannotated protein;sec_met_domain=HMGL-like (E-value: 1.2e-26, bitscore: 83.4, seeds: 15, tool: rule-based-clusters);source=Prodigal_v2.6.3;transl_table=11;\", 'test_contig_240684\\tantismash\\tprotocluster\\t1\\t11811\\t.\\t+\\t0\\taStool=rule-based-clusters;category=other;contig_edge=True;core_location=[0:11810];cutoff=10000;detection_rule=(HMGL-like and AMP-binding);neighbourhood=10000;product=betalactone;protocluster_number=1;tool=antismash;', 'test_contig_240684\\tantismash\\tproto_core\\t1\\t11810\\t.\\t+\\t0\\taStool=rule-based-clusters;tool=antismash;cutoff=10000;detection_rule=(HMGL-like and AMP-binding);neighbourhood=10000;product=betalactone;protocluster_number=1;', 'test_contig_240684\\tantismash\\tcand_cluster\\t1\\t11811\\t.\\t+\\t0\\tcandidate_cluster_number=1;contig_edge=True;detection_rules=(HMGL-like and AMP-binding);kind=single;product=betalactone;protoclusters=1;tool=antismash;', 'test_contig_240684\\tantismash\\tregion\\t1\\t11811\\t.\\t+\\t0\\tcandidate_cluster_numbers=1;contig_edge=True;product=betalactone;region_number=1;rules=(HMGL-like and AMP-binding);tool=antismash;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t625\\t927\\t.\\t-\\t0\\taSDomain=HMGL-like;aSTool=clusterhmmer;database=34.0;db_xref=PF00682.21;description=HMGL-like;detection=hmmscan;evalue=1.30E-23;label=JJJEIFBD_328428;locus_tag=JJJEIFBD_328428;protein_end=103;protein_start=2;score=84.0;tool=antismash;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t974\\t1129\\t.\\t+\\t0\\taSDomain=HMGL-like;aSTool=clusterhmmer;database=34.0;db_xref=PF00682.21;description=HMGL-like;detection=hmmscan;evalue=5.90E-11;label=JJJEIFBD_328429;locus_tag=JJJEIFBD_328429;protein_end=52;protein_start=0;score=42.5;tool=antismash;', \"test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t974\\t1948\\t.\\t+\\t0\\tgene_functions=['biosynthetic (rule-based-clusters) betalactone: HMGL-like', 'biosynthetic-additional (smcogs) SMCOG1271:2-isopropylmalate synthase (Score: 203.6; E-value: 9.5e-62)'];gene_kind=biosynthetic;inference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328429;phase=0;product=unannotated protein;sec_met_domain=HMGL-like (E-value: 4e-14, bitscore: 42.3, seeds: 15, tool: rule-based-clusters);source=Prodigal_v2.6.3;transl_table=11;\", 'test_contig_240684\\tantismash\\tPFAM_domain\\t1469\\t1894\\t.\\t+\\t0\\taSDomain=LeuA_dimer;aSTool=clusterhmmer;database=34.0;db_xref=PF08502.12;description=LeuA allosteric (dimerisation) domain;detection=hmmscan;evalue=7.80E-27;label=JJJEIFBD_328429;locus_tag=JJJEIFBD_328429;protein_end=307;protein_start=165;score=93.9;tool=antismash;', 'test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t2072\\t2989\\t.\\t+\\t0\\tinference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328430;phase=0;product=unannotated protein;source=Prodigal_v2.6.3;transl_table=11;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t2117\\t2527\\t.\\t+\\t0\\taSDomain=EamA;aSTool=clusterhmmer;database=34.0;db_xref=PF00892.22;description=EamA-like transporter family;detection=hmmscan;evalue=6.70E-08;label=JJJEIFBD_328430;locus_tag=JJJEIFBD_328430;protein_end=152;protein_start=15;score=32.9;tool=antismash;', 'test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t3011\\t3622\\t.\\t-\\t0\\tinference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328431;phase=0;product=unannotated protein;source=Prodigal_v2.6.3;transl_table=11;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t3083\\t3478\\t.\\t-\\t0\\taSDomain=Lysine_decarbox;aSTool=clusterhmmer;database=34.0;db_xref=PF03641.16;description=Possible lysine decarboxylase;detection=hmmscan;evalue=1.70E-39;label=JJJEIFBD_328431;locus_tag=JJJEIFBD_328431;protein_end=180;protein_start=48;score=135.0;tool=antismash;', 'test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t3736\\t5454\\t.\\t+\\t0\\tinference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328432;phase=0;product=unannotated protein;source=Prodigal_v2.6.3;transl_table=11;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t5233\\t5373\\t.\\t+\\t0\\taSDomain=LysM;aSTool=clusterhmmer;database=34.0;db_xref=PF01476.22;description=LysM domain;detection=hmmscan;evalue=1.60E-09;label=JJJEIFBD_328432;locus_tag=JJJEIFBD_328432;protein_end=546;protein_start=499;score=37.8;tool=antismash;', 'test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t5518\\t5841\\t.\\t+\\t0\\tgene_functions=regulatory (smcogs) SMCOG1058:ArsR family transcriptional regulator (Score: 67.3; E-value: 1.4e-20);gene_kind=regulatory;inference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328433;phase=0;product=unannotated protein;source=Prodigal_v2.6.3;transl_table=11;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t5596\\t5736\\t.\\t+\\t0\\taSDomain=HTH_5;aSTool=clusterhmmer;database=34.0;db_xref=PF01022.22;description=Bacterial regulatory protein, arsR family;detection=hmmscan;evalue=5.80E-14;label=JJJEIFBD_328433;locus_tag=JJJEIFBD_328433;protein_end=73;protein_start=26;score=51.8;tool=antismash;', 'test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t5883\\t7766\\t.\\t+\\t0\\tgene_functions=transport (smcogs) SMCOG1288:ABC transporter related protein (Score: 325.2; E-value: 1.4e-98);gene_kind=transport;inference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328434;phase=0;product=unannotated protein;source=Prodigal_v2.6.3;transl_table=11;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t5994\\t6827\\t.\\t+\\t0\\taSDomain=ABC_membrane;aSTool=clusterhmmer;database=34.0;db_xref=PF00664.25;description=ABC transporter transmembrane region;detection=hmmscan;evalue=1.00E-32;label=JJJEIFBD_328434;locus_tag=JJJEIFBD_328434;protein_end=315;protein_start=37;score=114.0;tool=antismash;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t7011\\t7460\\t.\\t+\\t0\\taSDomain=ABC_tran;aSTool=clusterhmmer;database=34.0;db_xref=PF00005.29;description=ABC transporter;detection=hmmscan;evalue=2.20E-33;label=JJJEIFBD_328434;locus_tag=JJJEIFBD_328434;protein_end=526;protein_start=376;score=115.8;tool=antismash;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t7858\\t8193\\t.\\t+\\t0\\taSDomain=YCII;aSTool=clusterhmmer;database=34.0;db_xref=PF03795.16;description=YCII-related domain;detection=hmmscan;evalue=5.50E-20;label=JJJEIFBD_328435;locus_tag=JJJEIFBD_328435;protein_end=112;protein_start=0;score=71.7;tool=antismash;', 'test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t7858\\t8229\\t.\\t+\\t0\\tinference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328435;phase=0;product=unannotated protein;source=Prodigal_v2.6.3;transl_table=11;', 'test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t8315\\t9010\\t.\\t+\\t0\\tinference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328436;phase=0;product=unannotated protein;source=Prodigal_v2.6.3;transl_table=11;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t8450\\t8965\\t.\\t+\\t0\\taSDomain=PS_Dcarbxylase;aSTool=clusterhmmer;database=34.0;db_xref=PF02666.17;description=Phosphatidylserine decarboxylase;detection=hmmscan;evalue=1.30E-31;label=JJJEIFBD_328436;locus_tag=JJJEIFBD_328436;protein_end=217;protein_start=45;score=109.9;tool=antismash;', 'test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t9019\\t9849\\t.\\t+\\t0\\tinference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328437;phase=0;product=unannotated protein;source=Prodigal_v2.6.3;transl_table=11;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t9097\\t9288\\t.\\t+\\t0\\taSDomain=CDP-OH_P_transf;aSTool=clusterhmmer;database=34.0;db_xref=PF01066.23;description=CDP-alcohol phosphatidyltransferase;detection=hmmscan;evalue=4.20E-12;label=JJJEIFBD_328437;locus_tag=JJJEIFBD_328437;protein_end=90;protein_start=26;score=46.7;tool=antismash;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t9652\\t9756\\t.\\t+\\t0\\taSDomain=CDP-OH_P_tran_2;aSTool=clusterhmmer;database=34.0;db_xref=PF08009.13;description=CDP-alcohol phosphatidyltransferase 2;detection=hmmscan;evalue=4.00E-09;label=JJJEIFBD_328437;locus_tag=JJJEIFBD_328437;protein_end=246;protein_start=211;score=36.3;tool=antismash;', \"test_contig_240684\\tProdigal_v2.6.3\\tCDS\\t9855\\t11810\\t.\\t-\\t0\\tNRPS_PKS=['Domain: AMP-binding (170-549). E-value: 2.6e-68. Score: 222.3. Matches aSDomain: nrpspksdomains_JJJEIFBD_328438_AMP-binding.1', 'type: NRPS-like protein'];gene_functions=['biosynthetic (rule-based-clusters) betalactone: AMP-binding', 'biosynthetic-additional (smcogs) SMCOG1002:AMP-dependent synthetase and ligase (Score: 265.3; E-value: 1.4e-80)'];gene_kind=biosynthetic;inference=ab initio prediction:Prodigal_v2.6.3;locus_tag=JJJEIFBD_328438;phase=0;product=unannotated protein;sec_met_domain=AMP-binding (E-value: 1.4e-72, bitscore: 234.0, seeds: 400, tool: rule-based-clusters);source=Prodigal_v2.6.3;transl_table=11;\", 'test_contig_240684\\tantismash\\taSDomain\\t10164\\t11300\\t.\\t-\\t0\\taSDomain=AMP-binding;aSTool=nrps_pks_domains;database=nrpspksdomains.hmm;detection=hmmscan;evalue=2.60E-68;label=JJJEIFBD_328438_AMP-binding.1;locus_tag=JJJEIFBD_328438;protein_end=549;protein_start=170;score=222.3;specificity=consensus: X;tool=antismash;', 'test_contig_240684\\tantismash\\tPFAM_domain\\t10164\\t11300\\t.\\t-\\t0\\taSDomain=AMP-binding;aSTool=clusterhmmer;database=34.0;db_xref=PF00501.30;description=AMP-binding enzyme;detection=hmmscan;evalue=2.70E-64;label=JJJEIFBD_328438;locus_tag=JJJEIFBD_328438;protein_end=549;protein_start=170;score=217.5;tool=antismash;', 'test_contig_240684\\tantismash\\tCDS_motif\\t10833\\t10889\\t.\\t-\\t0\\taSTool=nrps_pks_domains;database=abmotifs;detection=hmmscan;evalue=2.00E-06;label=NRPS-A_a3;locus_tag=JJJEIFBD_328438;protein_end=326;protein_start=307;score=20.4;tool=antismash;']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "lines = []\n",
    "line = ''\n",
    "for i in range(len(genome_record.features)):\n",
    "    #check if the type of region/feature is a coding sequence/annotated region with the 'phase' qualifier/key, to also include along the line\n",
    "    if genome_record.features[i].qualifiers.__contains__('phase'):\n",
    "        #check if annotation is from a 'tool' a 'aStool' or a 'source' to include in the second column \n",
    "        if genome_record.features[i].qualifiers.__contains__('tool'):\n",
    "            string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['tool'][0]) + '\\t' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start + 1) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' + str(genome_record.features[i].qualifiers['phase'][0]) + '\\t'\n",
    "        else:\n",
    "            if genome_record.features[i].qualifiers.__contains__('aStool'):\n",
    "                string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['aStool'][0]) + '\\t' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start + 1) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' + str(genome_record.features[i].qualifiers['phase'][0]) + '\\t'\n",
    "            else:\n",
    "                string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['source'][0]) + '\\t' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start + 1) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' + str(genome_record.features[i].qualifiers['phase'][0]) + '\\t'\n",
    "    else:\n",
    "        if genome_record.features[i].qualifiers.__contains__('tool'):\n",
    "            string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['tool'][0]) + '\\t' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start + 1) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' + str(0) + '\\t'\n",
    "        else:\n",
    "            if genome_record.features[i].qualifiers.__contains__('aStool'):\n",
    "                string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['aStool'][0]) + '\\t' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start + 1) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' + str(0) + '\\t'\n",
    "            else:\n",
    "                string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['source'][0]) + '\\t' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start + 1) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' + str(0) + '\\t'\n",
    "\n",
    "    attributess = ''\n",
    "    for key in genome_record.features[i].qualifiers:\n",
    "    #print(i, qualif_dict[i])\n",
    "        if key ==  'ID' or key == 'gene' or key == 'translation' or key == 'partial' or key == 'domain_id':\n",
    "            pass\n",
    "        else:\n",
    "            if len(genome_record.features[i].qualifiers[key]) > 1:\n",
    "                temp = str(key) + '=' + str(genome_record.features[i].qualifiers[key]) + ';' \n",
    "                attributess = attributess + str(temp)\n",
    "            else:\n",
    "            #print(len(genome_record.features[i].qualifiers[key]))\n",
    "                temp = str(key) + '=' + str(genome_record.features[i].qualifiers[key][0]) + ';' \n",
    "                attributess = attributess + str(temp)\n",
    "    \n",
    "    #attributess \n",
    "    line = line + string + attributess\n",
    "    lines.append(str(line))\n",
    "    line = ''\n",
    "    #print(line)\n",
    "#print(line)\n",
    "#lines.append(str(line))\n",
    "#line = ''\n",
    "print(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#function to get the highest scoring BGC term from the knownclusterblast text file for each of the found loci\n",
    "def get_bgc_terms(file, list):\n",
    "    #open file\n",
    "    f = open(file[0], 'r')\n",
    "\n",
    "\n",
    "    #dictionary to store the MiBig identifier with the corresponding locus_tag \n",
    "    locus_dict = {}    \n",
    "\n",
    "    for Line in f:\n",
    "        columns = Line.split('|')\n",
    "        locus = columns[4]\n",
    "        BGC_id = columns[5].split('\\t')[1]\n",
    "        #The highest is the first so if the key is already in the dictionary skip so no others are added for that locus tag\n",
    "        if locus in locus_dict.keys():\n",
    "            pass\n",
    "        else:\n",
    "            locus_dict[locus] = BGC_id\n",
    "    \n",
    "    file_list = list \n",
    "    for key, value in locus_dict.items():\n",
    "        for line in file_list:\n",
    "            if key in line:\n",
    "                #print(line)\n",
    "                line = line + value \n",
    "\n",
    "    f.close()\n",
    "    return file_list\n",
    "\n",
    "    \n",
    "\n",
    "final_list = get_bgc_terms(cb_txtfile, file_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"../OUTPUTS/antismash_finals/T_712_11/mg.assembly.merged.length_filtered/knownclusterblastoutput.txt\", 'r')\n",
    "\n",
    "\n",
    "#dictionary to store the MiBig identifier with the corresponding locus_tag \n",
    "locus_dict = {}    \n",
    "\n",
    "for Line in f:\n",
    "    columns = Line.split('|')\n",
    "    locus = columns[4]\n",
    "    BGC_id = columns[5].split('\\t')[1]\n",
    "    #The highest is the first so if the key is already in the dictionary skip so no others are added for that locus tag\n",
    "    if locus in locus_dict.keys():\n",
    "        pass\n",
    "    else:\n",
    "        locus_dict[locus] = BGC_id\n",
    "\n",
    "#file_list = list \n",
    "for key, value in locus_dict.items():\n",
    "    #print(key)\n",
    "    for line in lines:\n",
    "        if str(key) in line:\n",
    "            print(key)\n",
    "            line = line + value\n",
    "#print(lines)\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'JJJEIFBD_2894428': 'BGC0001751'}\n",
      "{'JJJEIFBD_2894428': 'BGC0001751', 'JJJEIFBD_2894431': 'BGC0002007'}\n",
      "{'JJJEIFBD_2894428': 'BGC0001751', 'JJJEIFBD_2894431': 'BGC0002007', 'JJJEIFBD_2894432': 'BGC0000268'}\n",
      "{'JJJEIFBD_2894428': 'BGC0001751', 'JJJEIFBD_2894431': 'BGC0002007', 'JJJEIFBD_2894432': 'BGC0000268', 'JJJEIFBD_2894434': 'BGC0001061'}\n",
      "{'JJJEIFBD_2894428': 'BGC0001751', 'JJJEIFBD_2894431': 'BGC0002007', 'JJJEIFBD_2894432': 'BGC0000268', 'JJJEIFBD_2894434': 'BGC0001061', 'JJJEIFBD_2894435': 'BGC0000817'}\n",
      "{'JJJEIFBD_2894428': 'BGC0001751', 'JJJEIFBD_2894431': 'BGC0002007', 'JJJEIFBD_2894432': 'BGC0000268', 'JJJEIFBD_2894434': 'BGC0001061', 'JJJEIFBD_2894435': 'BGC0000817', 'JJJEIFBD_2894436': 'BGC0001069'}\n",
      "{'JJJEIFBD_2894428': 'BGC0001751', 'JJJEIFBD_2894431': 'BGC0002007', 'JJJEIFBD_2894432': 'BGC0000268', 'JJJEIFBD_2894434': 'BGC0001061', 'JJJEIFBD_2894435': 'BGC0000817', 'JJJEIFBD_2894436': 'BGC0001069', 'JJJEIFBD_2894437': 'BGC0001520'}\n",
      "{'JJJEIFBD_2894428': 'BGC0001751', 'JJJEIFBD_2894431': 'BGC0002007', 'JJJEIFBD_2894432': 'BGC0000268', 'JJJEIFBD_2894434': 'BGC0001061', 'JJJEIFBD_2894435': 'BGC0000817', 'JJJEIFBD_2894436': 'BGC0001069', 'JJJEIFBD_2894437': 'BGC0001520', 'JJJEIFBD_2894438': 'BGC0001091'}\n",
      "{'JJJEIFBD_2894428': 'BGC0001751', 'JJJEIFBD_2894431': 'BGC0002007', 'JJJEIFBD_2894432': 'BGC0000268', 'JJJEIFBD_2894434': 'BGC0001061', 'JJJEIFBD_2894435': 'BGC0000817', 'JJJEIFBD_2894436': 'BGC0001069', 'JJJEIFBD_2894437': 'BGC0001520', 'JJJEIFBD_2894438': 'BGC0001091', 'JJJEIFBD_2894439': 'BGC0000180'}\n"
     ]
    }
   ],
   "source": [
    "#can add a for loop to find the highest scoring MiBig identifier for the locus tag \n",
    "f = open(\"../OUTPUTS/antismash_finals/T_712_11/mg.assembly.merged.length_filtered/knownclusterblastoutput.txt\", 'r')\n",
    "\n",
    "# if attributess.__contains__('locus_tag'):\n",
    "#     locus_tag = re.search(';locus_tag=(.+?);', attributess)\n",
    "#     locus_name = locus_tag.group(1)\n",
    "\n",
    "#dictionary to store the MiBig identifier with the corresponding locus_tag \n",
    "locus_dict = {}    \n",
    "\n",
    "for Line in f:\n",
    "    columns = Line.split('|')\n",
    "    locus = columns[4]\n",
    "    BGC_id = columns[5].split('\\t')[1]\n",
    "    #print(locus_name + locus)\n",
    "    if locus in locus_dict.keys():\n",
    "        pass\n",
    "    else:\n",
    "        locus_dict[locus] = BGC_id\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['rule-based-clusters']\n",
      "['RiPP']\n",
      "['True']\n",
      "['[5459:10452]']\n",
      "['20000']\n",
      "['((LANC_like and (Lant_dehydr_N or Lant_dehydr_C)) and not (YcaO or TIGR03882))']\n",
      "['10000']\n",
      "['lanthipeptide-class-i']\n",
      "['1']\n",
      "['antismash']\n",
      "['rule-based-clusters']\n",
      "['antismash']\n",
      "['20000']\n",
      "['((LANC_like and (Lant_dehydr_N or Lant_dehydr_C)) and not (YcaO or TIGR03882))']\n",
      "['10000']\n",
      "['lanthipeptide-class-i']\n",
      "['1']\n",
      "['1']\n",
      "['True']\n",
      "['((LANC_like and (Lant_dehydr_N or Lant_dehydr_C)) and not (YcaO or TIGR03882))']\n",
      "['single']\n",
      "['lanthipeptide-class-i']\n",
      "['1']\n",
      "['antismash']\n",
      "['1']\n",
      "['True']\n",
      "['lanthipeptide-class-i']\n",
      "['1']\n",
      "['((LANC_like and (Lant_dehydr_N or Lant_dehydr_C)) and not (YcaO or TIGR03882))']\n",
      "['antismash']\n",
      "['EMEGMFMJ_342191']\n",
      "['EMEGMFMJ_342191']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342191']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['MEPLPGIRIRRLRLVGMERSYDVDFTAEQRVRNLSVIAGAFSSGKTAVLEFIAYGLGAKRHPRHQEVLRRVRSCLMEVELSGEPHVIERSVGEPSKVAFVRRGTLQSRPTSRAESRTIDPPGAPESLSSLLLSHCKLEGVQLREAPTNSESRTDPLSFRDLMWLAFLPNERVADKNFLFENDYMRKHKLRQVVDVAFGVHDDRAVELGQRIQELAGRLSRAKAELAAARTFVLEQDPTALDGEPTPAVHEQELADITSRLTELDRRAQAGTEFAAQLRRQHKEAAQTSHRAAALLRDCETQLQRLMPLRAQYADDLLKLGMLGEAQQIFDPLSVTTCPACLNRLSAPPTANNGLCSLCRHELTAADGRLVLGTVDTGSQADENRVDVAAETRSTKARLKEITTYIDELGASLAALKLRAEEAAVQEEEAAAALDAATSPSVSPFLAARDDLHRRREQVLHQLEQAGSTVKLREGLTKRADAVERHEAQIARLREELARLGDASHDRDRVIANISSRYGELLRAWRYPKVSAPFIKTDLTPFVRGEPYQEASSGARTLLTLAWQLAVFEIAVEAGAAHPGFLMIDSPQKNLGHGGTLDAVIADAVAIDDFYHHLTLWLAERGSRAQLIVADNSPPPLVENSVVVRYSRNEDRPPYGLIEDETSADDEADAEM']\n",
      "['EMEGMFMJ_342192']\n",
      "['EMEGMFMJ_342192']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342192']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['VRTDRPVMMPEDEVPFRLAQLLLLLDAVAAQDANGATLERIGYYDFLSANPFLVVPSEGRDASLLRLAGFDPQVLAYASSSQRFTSRRERIQHDLALLVAYGCCRVRNRNGSLTYLITEAGQHLGREFTATYATSFATAADIVVRQLRKLSDKRLREQTALWLKPDGPVAALMSVLGPGPLLETSWEG']\n",
      "['EMEGMFMJ_342193']\n",
      "['EMEGMFMJ_342193']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342193']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['LDNSKPVAAGVPPIHFVVHELRTGSADGARADFEQMLAQLASATNPNVRMIAANPGDWGIDAFAGDLGGAITVWQSKYFMPKTTTSQSQQIRESLANGLKAAAANGHTITLWILCIPSSMDGPAAQWWDGWKRRQEKAHGLVIELWDETKIKKKLLSPEGDQVRRAYYEPFTAPAPAQEQIRLVLDVEEDKAAALDSALFVRQMTEAGHVELDSAKRQFFNADLVAREVAHKGVPGEVAALSSADATLHGVWEMQFNECSAEGTLNALYGRVWREVRGEHDKLPKVLRLEVTHSWGLVHRLVDNRRAGWVAHWRQIASGHPDD']\n",
      "['EMEGMFMJ_342194']\n",
      "['EMEGMFMJ_342194']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342194']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['VWKLPDTSAQRFPLVARPRPACLPLPERVQALADLADTAAKQNDPGVASTVYNQAALIASDVGLPDIARSICTKHAELYLRAAPLSHETAVRALEPVVNLARLHLRAGRQDTARQYLSMLFDAVTHAQPAQIESLTIPADLVATDDDRKEVRTWLWTVLLADGTRSLTSAGRWNDALNHIQQHHGLAQRMLDGRQVAVLAALASGDNAHAQEVLAHTKPGEPWEAAVTECLTVMSLRGAEEEWRRPLRQLVTVYLDLPDVEGLTVFTTRYGLAVLDLIGTTDDNAARLVVADLHRRAVAANDGYAARDALSHPVSRSLASDQEFKESQRLVNACTLRSGPFQLASWGRIDGVLHVSRRVIQESFNEGAL']\n",
      "['EMEGMFMJ_342195']\n",
      "['EMEGMFMJ_342195']\n",
      "['biosynthetic (rule-based-clusters) lanthipeptide-class-i: Lant_dehydr_C']\n",
      "['biosynthetic']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342195']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['Lant_dehydr_C (E-value: 4.5e-38, bitscore: 121.1, seeds: 35, tool: rule-based-clusters)']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['MPPPTWRQVNVAFSDWERAEATAMTELVPLVEAAEDAGAITRWFFVRKHPCWRIRYASVAGTQDAIDDRLDALVAAGSTAGWTESVYEPEVHAFGGTQAMDSAHRLFHQDSHNLLHFLRGDTARHRRETSLMLCSLMLRAASLDWYEQGDVWARVAAHRPPSPDAEPDNKDRLHTAVRRLITVDPQYATQPGSPLAGAEAWTAAYTAAGSDLARLNQTGQLHRGIRDVLAHHVIFAWNRLGLPYSTQAALTTAAQTVIFGPDPSTERSTAHRVETP']\n",
      "['Lant_dehydr_C']\n",
      "['clusterhmmer']\n",
      "['34.0']\n",
      "['PF14028.8']\n",
      "['Lantibiotic biosynthesis dehydratase C-term']\n",
      "['hmmscan']\n",
      "['clusterhmmer_EMEGMFMJ_342195_0001']\n",
      "['6.70E-39']\n",
      "['EMEGMFMJ_342195']\n",
      "['EMEGMFMJ_342195']\n",
      "['256']\n",
      "['5']\n",
      "['134.4']\n",
      "['antismash']\n",
      "['WRQVNVAFSDWERAEATAMTELVPLVEAAEDAGAITRWFFVRKHPCWRIRYASVAGTQDAIDDRLDALVAAGSTAGWTESVYEPEVHAFGGTQAMDSAHRLFHQDSHNLLHFLRGDTARHRRETSLMLCSLMLRAASLDWYEQGDVWARVAAHRPPSPDAEPDNKDRLHTAVRRLITVDPQYATQPGSPLAGAEAWTAAYTAAGSDLARLNQTGQLHRGIRDVLAHHVIFAWNRLGLPYSTQAALTTAAQT']\n",
      "['EMEGMFMJ_342196']\n",
      "['EMEGMFMJ_342196']\n",
      "['biosynthetic (rule-based-clusters) lanthipeptide-class-i: LANC_like', 'biosynthetic-additional (smcogs) SMCOG1140:Lanthionine synthetase C family protein (Score: 365.1; E-value: 7.1e-111)']\n",
      "['biosynthetic']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342196']\n",
      "['smCOG tree PNG image: smcogs/EMEGMFMJ_342196.png']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['LANC_like (E-value: 4.4e-44, bitscore: 140.3, seeds: 47, tool: rule-based-clusters)']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['VTTALTSAVLSTTDKIAKALRDPSSAWPNGRPAGGRAWPQSLAGGACGIALLHIERARTGHGDWDTAHTWLAEAVGGEISAADNANLYFGASALAYITHRAASSSSRYLPLLNHLDTATLTVTRIRIAAAHQRIDRAEPVPMEEFDLIQGLSGLGAYHLSRHPHHEITRDTLTYLVRLTEPLSVPTGLPPWWMDVAPTGAPHPDYPDGHGNFGLAHGIGAALAVLSQALLHEVTVPGMTDAVARICAWTDRWRQGDDNHPWWPGHITARQAADRHVHADLRPRPSWCYGASGTARAQQLAGLALGDPARVQLAENAILDALRDPDQLSKLREIGLCHGMAGLLHAGWRMATETGNPEITAELPHLAERLIAALDQDTQDPELLDGSAGAALALHTYATAVAPAPHWDTFLSLA']\n",
      "['LANC_like']\n",
      "['clusterhmmer']\n",
      "['34.0']\n",
      "['PF05147.15']\n",
      "['Lanthionine synthetase C-like protein']\n",
      "['hmmscan']\n",
      "['clusterhmmer_EMEGMFMJ_342196_0002']\n",
      "['1.20E-42']\n",
      "['EMEGMFMJ_342196']\n",
      "['EMEGMFMJ_342196']\n",
      "['406']\n",
      "['37']\n",
      "['145.9']\n",
      "['antismash']\n",
      "['WPQSLAGGACGIALLHIERARTGHGDWDTAHTWLAEAVGGEISAADNANLYFGASALAYITHRAASSSSRYLPLLNHLDTATLTVTRIRIAAAHQRIDRAEPVPMEEFDLIQGLSGLGAYHLSRHPHHEITRDTLTYLVRLTEPLSVPTGLPPWWMDVAPTGAPHPDYPDGHGNFGLAHGIGAALAVLSQALLHEVTVPGMTDAVARICAWTDRWRQGDDNHPWWPGHITARQAADRHVHADLRPRPSWCYGASGTARAQQLAGLALGDPARVQLAENAILDALRDPDQLSKLREIGLCHGMAGLLHAGWRMATETGNPEITAELPHLAERLIAALDQDTQDPELLDGSAGAALALHTYATAVAPAPHW']\n",
      "['EMEGMFMJ_342197']\n",
      "['EMEGMFMJ_342197']\n",
      "['biosynthetic (rule-based-clusters) lanthipeptide-class-i: Lant_dehydr_C', 'biosynthetic (rule-based-clusters) lanthipeptide-class-i: Lant_dehydr_N', 'biosynthetic-additional (rule-based-clusters) Lanthipeptide_LanB_RRE', 'biosynthetic-additional (smcogs) SMCOG1155:Lantibiotic dehydratase domain protein (Score: 856.3; E-value: 4.7e-259)']\n",
      "['biosynthetic']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342197']\n",
      "['smCOG tree PNG image: smcogs/EMEGMFMJ_342197.png']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['Lant_dehydr_N (E-value: 4e-113, bitscore: 368.3, seeds: 73, tool: rule-based-clusters)', 'Lant_dehydr_C (E-value: 3.2e-29, bitscore: 92.1, seeds: 35, tool: rule-based-clusters)', 'Lanthipeptide_LanB_RRE (E-value: 4.8e-21, bitscore: 65.1, seeds: 20, tool: rule-based-clusters)']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['MYRYVDAALIRAAAGQPDTQVTWPDLVSTTAGAVSWCAWLQEVWRIGDFADAVSAASPDLASRVEQICTGHHLPEPTVRRIVLAVLRYLLRARTRATPFGLFAGVAAASIGAAPALRLGSEHRPTARADAARSTSLIDRFEQHPALRPHLMLLTSTLAVERDDQVVIDHRPGGKPGAGLEHVHIRMTEPVRAALDGAQSPIPWCDLASKLAAGFPDTSPNLIDKLLASLVTQQFLLTSLRPAMATHDPLTGLHDHEHRLSNQERAEVRATPRHAQDLLLDCDLTIPDTVAREATAAAKALTRLAPRAALTGWAEWHDRFLDRYGPRAVVPVLDAIDALGYPPGYLGATAAPTLSPLPDRDARLIKLAHTAAVQGRLEVVLDDAALESLATTGAEHPVQPSTELTVRIDAADSAALEQGEFSLHVVGVARSAGATTGRFLDMLPDEHRRRMTEVYAALPGVHDDALVAQISAVPLSVRAANVARAPQVTDLVISLGEYRGPGSRLLPVTDLAVIADAGQLHLVSLSRRRPVHALLLNAVDLGLHSHPLTRFLAEAPVALAVPCAGFLWGTAASNLPFLPALRYGRTILSPARWLLARGDLPAASTPWLQWDEALSRWCRDVRLPERVYLSEADQCLALDLTEPSHRALLRTHLDRDGTVTLRPAPEPQDLGWIGGRAHEVVIPLISDQNHTSVRTAPHVVTREHGRLPGSGELLYVQLHGDPARQDHILTRHLPVLLAELKPPRCWFVRYSEPGHHLRLRLSCTPGALGATFEYVGAWTEELRRHGLITHVSVETYHPEIARYGGPTAIDAAEAYFAADSTAVLAQLVAGLAKNAPDARAVTVASMVDIATGLLGDRARAMGWLIAHTRTDSAAPPRPVHRQAVELVHTHPVGLAEQTAGAWSARRMALADYAEALLAAGADPGDLLPDLLHLHHVRMHGPGLPEERTHLHLARAAALSWTARSRRTA']\n",
      "['Lant_dehydr_C']\n",
      "['clusterhmmer']\n",
      "['34.0']\n",
      "['PF14028.8']\n",
      "['Lantibiotic biosynthesis dehydratase C-term']\n",
      "['hmmscan']\n",
      "['clusterhmmer_EMEGMFMJ_342197_0005']\n",
      "['8.80E-28']\n",
      "['EMEGMFMJ_342197']\n",
      "['EMEGMFMJ_342197']\n",
      "['956']\n",
      "['712']\n",
      "['98.0']\n",
      "['antismash']\n",
      "['LYVQLHGDPARQDHILTRHLPVLLAELKPPRCWFVRYSEPGHHLRLRLSCTPGALGATFEYVGAWTEELRRHGLITHVSVETYHPEIARYGGPTAIDAAEAYFAADSTAVLAQLVAGLAKNAPDARAVTVASMVDIATGLLGDRARAMGWLIAHTRTDSAAPPRPVHRQAVELVHTHPVGLAEQTAGAWSARRMALADYAEALLAAGADPGDLLPDLLHLHHVRMHGPGLPEERTHLHLARAAA']\n",
      "['Lant_dehydr_N']\n",
      "['clusterhmmer']\n",
      "['34.0']\n",
      "['PF04738.15']\n",
      "['Lantibiotic dehydratase, N terminus']\n",
      "['hmmscan']\n",
      "['clusterhmmer_EMEGMFMJ_342197_0004']\n",
      "['1.80E-112']\n",
      "['EMEGMFMJ_342197']\n",
      "['EMEGMFMJ_342197']\n",
      "['648']\n",
      "['270']\n",
      "['376.6']\n",
      "['antismash']\n",
      "['PRHAQDLLLDCDLTIPDTVAREATAAAKALTRLAPRAALTGWAEWHDRFLDRYGPRAVVPVLDAIDALGYPPGYLGATAAPTLSPLPDRDARLIKLAHTAAVQGRLEVVLDDAALESLATTGAEHPVQPSTELTVRIDAADSAALEQGEFSLHVVGVARSAGATTGRFLDMLPDEHRRRMTEVYAALPGVHDDALVAQISAVPLSVRAANVARAPQVTDLVISLGEYRGPGSRLLPVTDLAVIADAGQLHLVSLSRRRPVHALLLNAVDLGLHSHPLTRFLAEAPVALAVPCAGFLWGTAASNLPFLPALRYGRTILSPARWLLARGDLPAASTPWLQWDEALSRWCRDVRLPERVYLSEADQCLALDLTEPSHRALL']\n",
      "['Lant_dehydr_N']\n",
      "['clusterhmmer']\n",
      "['34.0']\n",
      "['PF04738.15']\n",
      "['Lantibiotic dehydratase, N terminus']\n",
      "['hmmscan']\n",
      "['clusterhmmer_EMEGMFMJ_342197_0003']\n",
      "['9.90E-56']\n",
      "['EMEGMFMJ_342197']\n",
      "['EMEGMFMJ_342197']\n",
      "['254']\n",
      "['48']\n",
      "['189.1']\n",
      "['antismash']\n",
      "['FADAVSAASPDLASRVEQICTGHHLPEPTVRRIVLAVLRYLLRARTRATPFGLFAGVAAASIGAAPALRLGSEHRPTARADAARSTSLIDRFEQHPALRPHLMLLTSTLAVERDDQVVIDHRPGGKPGAGLEHVHIRMTEPVRAALDGAQSPIPWCDLASKLAAGFPDTSPNLIDKLLASLVTQQFLLTSLRPAMATHDPLTGLHD']\n",
      "['tta leucine codon, possible target for bldA regulation']\n",
      "['antismash']\n",
      "['EMEGMFMJ_342198']\n",
      "['EMEGMFMJ_342198']\n",
      "['biosynthetic-additional (rule-based-clusters) LD_lanti_pre', 'biosynthetic-additional (lanthipeptides) predicted lanthipeptide']\n",
      "['biosynthetic-additional']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342198']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['LD_lanti_pre (E-value: 8.5e-19, bitscore: 57.0, seeds: 21, tool: rule-based-clusters)']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['MTVQMERPVTPPQPGDALAPTDAAGDWDLDITIVEGGPAADQLIRLTSDGCNSTCATACVSCP']\n",
      "['leader']\n",
      "['peptide class: lanthipeptide', 'predicted sequence: MTVQMERPVTPPQPGDALAPTDAAGDWDLDITIVEGGPAAD']\n",
      "['EMEGMFMJ_342198_lanthipeptide']\n",
      "['lanthipeptides']\n",
      "['antismash']\n",
      "['core']\n",
      "['EMEGMFMJ_342198_lanthipeptide']\n",
      "['lanthipeptide']\n",
      "['Class I']\n",
      "['-7.50']\n",
      "['2135.4']\n",
      "['2134.0']\n",
      "['QLIRLTSDGCNSTCATACVSCP']\n",
      "['lanthipeptides']\n",
      "['antismash']\n",
      "['MTVQMERPVTPPQPGDALAPTDAAGDWDLDITIVEGGPAAD ']\n",
      "['[10608:10731](-)']\n",
      "['2153.5', '2171.5']\n",
      "['EMEGMFMJ_342199']\n",
      "['EMEGMFMJ_342199']\n",
      "['biosynthetic-additional (rule-based-clusters) PCMT']\n",
      "['biosynthetic-additional']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342199']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['PCMT (E-value: 3.3e-39, bitscore: 124.5, seeds: 9, tool: rule-based-clusters)']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['MIAIDGSTDRDPDVLRADMVRALRDEDGITTGPVAAAFATVPRHLFAPDAPPEVAYSLHRTVPVKKDEHGMDISVMSAAHLQAVMLEQAEVEPGMNVLEIGSGGVNAAYLQELVGSEGRVTTVDIDGDVIDRARRCLDDAGYSRVRTVLDDGEHGVADGAPFDRIIVTAAAWDIPPSWIGGLAEKGRLVVPLTVCGTTRSIAFDRDGSGLVSRSYGLAHFVPMRGEGAADERKVMLREGVALQTDDLRVPLEPQALDAALDMPRLERWSGAAYDLPDELDLFITLNLPSPARLHAAQDVVGSGLVEASALKGVSALVSRDSFAYRTRRENAATGGFESGVVAHGPQAEALADQYADLLRHWAQNHRRRGAATFRYLPGPAPKTLPEGAVLRRHGIVTAVWT']\n",
      "['PCMT']\n",
      "['clusterhmmer']\n",
      "['34.0']\n",
      "['PF01135.21']\n",
      "['Protein-L-isoaspartate(D-aspartate) O-methyltransferase (PCMT)']\n",
      "['hmmscan']\n",
      "['clusterhmmer_EMEGMFMJ_342199_0006']\n",
      "['5.20E-36']\n",
      "['EMEGMFMJ_342199']\n",
      "['EMEGMFMJ_342199']\n",
      "['223']\n",
      "['17']\n",
      "['124.5']\n",
      "['antismash']\n",
      "['DMVRALRDEDGITTGPVAAAFATVPRHLFAPDAPPEVAYSLHRTVPVKKDEHGMDISVMSAAHLQAVMLEQAEVEPGMNVLEIGSGGVNAAYLQELVGSEGRVTTVDIDGDVIDRARRCLDDAGYSRVRTVLDDGEHGVADGAPFDRIIVTAAAWDIPPSWIGGLAEKGRLVVPLTVCGTTRSIAFDRDGSGLVSRSYGLAHFVPM']\n",
      "['EMEGMFMJ_342200']\n",
      "['EMEGMFMJ_342200']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342200']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['MTNTTVEPDEATRLRNDVVDTLRAEGWISSSRS']\n",
      "['EMEGMFMJ_342201']\n",
      "['EMEGMFMJ_342201']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342201']\n",
      "['00']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['LTTTSLALKPDTVAPPAQQPLPEGAASLGRTWHISTAGGITVTGYLPPWADEDPSESDVPLEELPVKLVDLAHWRPFDGQMMRIHNPAHADGADLVAEASLFDGHVTCHPYSDDPQERTPYADVRVVDDLWLYHLTPEDLTRLTAQLRAQADHIENDVTPLLEAARIDWTARQSGRTDVVVDVPVGASIASAPIGAVRC']\n",
      "['EMEGMFMJ_342202']\n",
      "['EMEGMFMJ_342202']\n",
      "['ab initio prediction:Prodigal_v2.6.3']\n",
      "['EMEGMFMJ_342202']\n",
      "['01']\n",
      "['0']\n",
      "['unannotated protein']\n",
      "['Prodigal_v2.6.3']\n",
      "['11']\n",
      "['QAHGGGDHFALVLDLDGVATGAIWTRWADGTPAALDLRPDCPFVDPESREGCCEFAFHPGAHSHCLTATPTEFS']\n",
      "[SeqFeature(FeatureLocation(ExactPosition(0), ExactPosition(13115), strand=1), type='protocluster'), SeqFeature(FeatureLocation(ExactPosition(5459), ExactPosition(10452), strand=1), type='proto_core'), SeqFeature(FeatureLocation(ExactPosition(0), ExactPosition(13115), strand=1), type='cand_cluster'), SeqFeature(FeatureLocation(ExactPosition(0), ExactPosition(13115), strand=1), type='region'), SeqFeature(FeatureLocation(ExactPosition(214), ExactPosition(2230), strand=-1), type='CDS')]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for i in range(len(genome_record.features)):\n",
    "    for key in genome_record.features[i].qualifiers:\n",
    "        print(genome_record.features[i].qualifiers[key])\n",
    "    #if genome_record.features[i].type == 'CDS':\n",
    "    #    print(genome_record.features[i])\n",
    "print(genome_record.features[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "antismash PFAM_domain\t0\t309\n"
     ]
    }
   ],
   "source": [
    "print(str(genome_record.features[0].qualifiers['tool'][0]) + ' ' + str(genome_record.features[0].type) + '\\t' + str(genome_record.features[0].location.start) + '\\t' + str(genome_record.features[0].location.end))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "#print(genome_record.features[0].qualifiers.__contains__('aStool'))\n",
    "print(str(genome_record.features[0].location.start + 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__add__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__nonzero__', '__radd__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_end', '_flip', '_get_strand', '_set_strand', '_shift', '_start', '_strand', 'end', 'extract', 'nofuzzy_end', 'nofuzzy_start', 'parts', 'ref', 'ref_db', 'start', 'strand']\n"
     ]
    }
   ],
   "source": [
    "print(dir(genome_record.features[0].location))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_functions=biosynthetic (rule-based-clusters) lanthipeptide-class-i: Lant_dehydr_C;gene_kind=biosynthetic;inference=ab initio prediction:Prodigal_v2.6.3;locus_tag=EMEGMFMJ_342197;note=smCOG tree PNG image: smcogs/EMEGMFMJ_342197.png;phase=0;product=unannotated protein;sec_met_domain=Lant_dehydr_N (E-value: 4e-113, bitscore: 368.3, seeds: 73, tool: rule-based-clusters);source=Prodigal_v2.6.3;transl_table=11;\n"
     ]
    }
   ],
   "source": [
    "attributess = ''\n",
    "#qualif_dict = genome_record.features[12].qualifiers\n",
    "for i in genome_record.features[12].qualifiers:\n",
    "#for i in qualif_dict:\n",
    "    #print(i, qualif_dict[i])\n",
    "    \n",
    "    if i ==  'ID' or i == 'gene' or i == 'translation' or i == 'partial' or i == 'domain_id':\n",
    "        pass\n",
    "    else:\n",
    "        temp = str(i) + '=' + str(genome_record.features[12].qualifiers[i][0]) + ';'\n",
    "        #temp = str(i) + '=' + str(qualif_dict[i][0]) + ';' \n",
    "        attributess = attributess + str(temp)\n",
    "print(attributess)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: CDS\n",
      "location: [6310:7552](-)\n",
      "qualifiers:\n",
      "    Key: ID, Value: ['EMEGMFMJ_342196']\n",
      "    Key: gene, Value: ['EMEGMFMJ_342196']\n",
      "    Key: gene_functions, Value: ['biosynthetic (rule-based-clusters) lanthipeptide-class-i: LANC_like', 'biosynthetic-additional (smcogs) SMCOG1140:Lanthionine synthetase C family protein (Score: 365.1; E-value: 7.1e-111)']\n",
      "    Key: gene_kind, Value: ['biosynthetic']\n",
      "    Key: inference, Value: ['ab initio prediction:Prodigal_v2.6.3']\n",
      "    Key: locus_tag, Value: ['EMEGMFMJ_342196']\n",
      "    Key: note, Value: ['smCOG tree PNG image: smcogs/EMEGMFMJ_342196.png']\n",
      "    Key: partial, Value: ['00']\n",
      "    Key: phase, Value: ['0']\n",
      "    Key: product, Value: ['unannotated protein']\n",
      "    Key: sec_met_domain, Value: ['LANC_like (E-value: 4.4e-44, bitscore: 140.3, seeds: 47, tool: rule-based-clusters)']\n",
      "    Key: source, Value: ['Prodigal_v2.6.3']\n",
      "    Key: transl_table, Value: ['11']\n",
      "    Key: translation, Value: ['VTTALTSAVLSTTDKIAKALRDPSSAWPNGRPAGGRAWPQSLAGGACGIALLHIERARTGHGDWDTAHTWLAEAVGGEISAADNANLYFGASALAYITHRAASSSSRYLPLLNHLDTATLTVTRIRIAAAHQRIDRAEPVPMEEFDLIQGLSGLGAYHLSRHPHHEITRDTLTYLVRLTEPLSVPTGLPPWWMDVAPTGAPHPDYPDGHGNFGLAHGIGAALAVLSQALLHEVTVPGMTDAVARICAWTDRWRQGDDNHPWWPGHITARQAADRHVHADLRPRPSWCYGASGTARAQQLAGLALGDPARVQLAENAILDALRDPDQLSKLREIGLCHGMAGLLHAGWRMATETGNPEITAELPHLAERLIAALDQDTQDPELLDGSAGAALALHTYATAVAPAPHWDTFLSLA']\n",
      "\n",
      "<class 'list'>\n",
      "['__add__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']\n",
      "['__add__', '__bool__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__le___', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__radd__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_per_letter_annotations', '_seq', '_set_per_letter_annotations', '_set_seq', 'annotations', 'dbxrefs', 'description', 'features', 'format', 'id', 'letter_annotations', 'lower', 'name', 'reverse_complement', 'seq', 'translate', 'upper']\n"
     ]
    }
   ],
   "source": [
    "#print(genome_record)\n",
    "#print(genome_record.features)\n",
    "#print(dir(genome_record))\n",
    "print(genome_record.features[10])\n",
    "print(type(genome_record.features))\n",
    "print(dir(genome_record.features))\n",
    "print(dir(genome_record))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test_contig_245285\tstr\n"
     ]
    }
   ],
   "source": [
    "print(genome_record.description + '\\t' + 'str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protocluster\n",
      "proto_core\n",
      "cand_cluster\n",
      "region\n",
      "CDS\n",
      "CDS\n",
      "CDS\n",
      "CDS\n",
      "CDS\n",
      "PFAM_domain\n",
      "CDS\n",
      "PFAM_domain\n",
      "CDS\n",
      "PFAM_domain\n",
      "PFAM_domain\n",
      "PFAM_domain\n",
      "misc_feature\n",
      "CDS\n",
      "CDS_motif\n",
      "CDS_motif\n",
      "CDS\n",
      "PFAM_domain\n",
      "CDS\n",
      "CDS\n",
      "CDS\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for i in genome_record.features:\n",
    "    #print(i.type)\n",
    "    print(i.type) #print(i.location)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write gff outfile with all the types entrees for antismash annotation from gbk file \n",
    "with open(r'./mg.all.gff', 'w') as out:\n",
    "    #for line in lines: \n",
    "    #    out.write('\\n' % line)\n",
    "    out.write('\\n'.join(lines))\n",
    "\n",
    "#write gff outfile for CDS regions which have actual \n",
    "with open(r'./mg.CDS.counts.gff', 'w') as CDS_out:\n",
    "    for line in lines:\n",
    "        if line.__contains__('gene_functions'):\n",
    "            CDS_out.write(line + '\\n')\n",
    "\n",
    "with open(r'./mg.Pfam.counts.gff', 'w') as Pfam_out:\n",
    "    for line in lines:\n",
    "        if line.__contains__('PFAM_domain'):\n",
    "            line = line.replace('PFAM_domain', 'CDS')\n",
    "            Pfam_out.write(line + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']\n"
     ]
    }
   ],
   "source": [
    "print(dir(str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_lines(file_path):\n",
    "    genome_record = SeqIO.read(file_path, 'genbank')\n",
    "    \n",
    "    lines = []\n",
    "    line = ''\n",
    "    for i in range(len(genome_record.features)):\n",
    "        #check if the type of region/feature is a coding sequence/annotated region with the 'phase' qualifier/key, to also include along the line\n",
    "        if genome_record.features[i].qualifiers.__contains__('phase'):\n",
    "            #check if annotation is from a 'tool' a 'aStool' or a 'source' to include in the second column \n",
    "            if genome_record.features[i].qualifiers.__contains__('tool'):\n",
    "                string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['tool'][0]) + ' ' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' + str(genome_record.features[i].qualifiers['phase'][0]) + '\\t'\n",
    "            else:\n",
    "                if genome_record.features[i].qualifiers.__contains__('aStool'):\n",
    "                    string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['aStool'][0]) + ' ' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' + str(genome_record.features[i].qualifiers['phase'][0]) + '\\t'\n",
    "                else:\n",
    "                    string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['source'][0]) + ' ' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' + str(genome_record.features[i].qualifiers['phase'][0]) + '\\t'\n",
    "        else:\n",
    "            if genome_record.features[i].qualifiers.__contains__('tool'):\n",
    "                string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['tool'][0]) + ' ' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t' \n",
    "            else:\n",
    "                if genome_record.features[i].qualifiers.__contains__('aStool'):\n",
    "                    string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['aStool'][0]) + ' ' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t'\n",
    "                else:\n",
    "                    string = str(genome_record.description) + '\\t' + str(genome_record.features[i].qualifiers['source'][0]) + ' ' + str(genome_record.features[i].type) + '\\t' + str(genome_record.features[i].location.start) + '\\t' + str(genome_record.features[i].location.end) + '\\t' + '.' + '\\t' + str(genome_record.features[i].location)[-2] + '\\t'\n",
    "        \n",
    "        attributess = ''\n",
    "        for key in genome_record.features[i].qualifiers:\n",
    "        #print(i, qualif_dict[i])\n",
    "            if key ==  'ID' or key == 'gene' or key == 'translation' or key == 'partial' or key == 'domain_id':\n",
    "                pass\n",
    "            else:\n",
    "                if len(genome_record.features[i].qualifiers[key]) > 1:\n",
    "                    temp = str(key) + '=' + str(genome_record.features[i].qualifiers[key]) + ';' \n",
    "                    attributess = attributess + str(temp)\n",
    "                else:\n",
    "                #print(len(genome_record.features[i].qualifiers[key]))\n",
    "                    temp = str(key) + '=' + str(genome_record.features[i].qualifiers[key][0]) + ';' \n",
    "                    attributess = attributess + str(temp)\n",
    "        #attributess \n",
    "        line = line + string + attributess\n",
    "        lines.append(str(line))\n",
    "        line = ''\n",
    "    #return list of all the lines \n",
    "    return lines \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob \n",
    "filenames = glob.glob('../OUTPUTS/IMP3_assembly/T_722_5/antismash_output' + \"/test_contig_*.gbk\") \n",
    "\n",
    "#loop over all the gbk files in the list and apply the above function to all\n",
    "file_list = []\n",
    "#for each file(by calling the name) in the list of file names\n",
    "for file in filenames:\n",
    "    #\n",
    "    tmp_file = make_lines(file)\n",
    "    file_list = file_list + tmp_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write gff outfile with all the types entrees for antismash annotation from gbk file \n",
    "with open(r'./mg.all.gff', 'w') as out:\n",
    "    #for line in lines: \n",
    "    #    out.write('\\n' % line)\n",
    "    out.write('\\n'.join(file_list))\n",
    "\n",
    "#write gff outfile for CDS regions which have actual annotated locus tags with identified gene_functions\n",
    "with open(r'./mg.CDS.counts.gff', 'w') as CDS_out:\n",
    "    for line in file_list:\n",
    "        if line.__contains__('gene_functions'):\n",
    "            CDS_out.write(line + '\\n')\n",
    "\n",
    "#write gff outfile for found regions from Pfam database \n",
    "with open(r'./mg.Pfam.counts.gff', 'w') as Pfam_out:\n",
    "    for line in file_list:\n",
    "        if line.__contains__('PFAM_domain'):\n",
    "            Pfam_out.write(line + '\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.5 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "0dd84e6e4655a1be09e25480d2aa07830f784093fb8ac0847c23715d011b335f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
