---
title: "testing"
author: "Susana Lopez Lemarroy"
date: "2022-09-30"
output: html_document
---

```{r}
#to use script that reads gff files 
##source("/Users/SusanaLopezLemarroy/Documents/Vrije_University_in_Amsterdam/INTERNSHIPs/Minor/tomato-rhizosphere/read.gff.R", local = ##knitr::knit_global())
```

```{r}
#libraries
##library(ggplot2)

#read files in folder to make functional profile data frame with mapping read counts for Pfam databse KOs 
##file.names <- list.files("../Pfam_output", pattern="*.tsv", full.names=TRUE)
##ldf <- lapply(file.names, read.table, header=TRUE)
##head(ldf)
#res <- lapply(ldf,)
#names(res) <- substr(file.names, 6, 30)
#res

##T_722_5 <- read.table('/Users/SusanaLopezLemarroy/Documents/Vrije_University_in_Amsterdam/INTERNSHIPs/Minor/OUTPUTS/IMP3_assembly/T_722_5/mg.Pfam_A.counts.tsv', header = TRUE)
##head(T_722_5)

#make a data frame with all the samples from the same tomato line to normalize the mapped counts based on average 
#fill the unmatched genes with NAs or 0s 
#data from mapped reads after annotation in IMP3 
##Pfam_map_profile <- T_722_5[,c("Geneid", "Assembly.mg.reads.sorted.bam")]
##head(Pfam_map_profile)

#how to compare to total number of reads, to observe what percentage of the reads mapped back to the annotated assembly. (thesis figure?)
```


